bindings/src/main/java/io/pravega/storage/extendeds3/ExtendedS3Storage.java

```diff
@@ -181,6 +185,71 @@ public boolean supportsTruncation() {
         return false;
     }
 
+    @Override
+    public Iterator<SegmentProperties> listSegments() {
+        return new ExtendedS3SegmentIterator(s3object -> true);
+    }
+
+    /**
+     * Iterator for segments in ExtendedS3Storage.
+     */
+    public class ExtendedS3SegmentIterator implements Iterator<SegmentProperties> {
+        ListObjectsResult results;
+        S3Object current;
+        Iterator<S3Object> innerIterator;
+        java.util.function.Predicate<S3Object> patternMatchPredicate;
+
+        ExtendedS3SegmentIterator(java.util.function.Predicate<S3Object> patternMatchPredicate) {
+            this.results = client.listObjects(config.getBucket(), config.getPrefix());
+            this.innerIterator = results.getObjects().iterator();
+            this.patternMatchPredicate = patternMatchPredicate;
+        }
+
+        @Override
+        public boolean hasNext() {
+            boolean nextBatch = false;
+            while (innerIterator != null) { // Loops through the batches
+                while (innerIterator.hasNext()) { // Loops through the objects in the batch
+                    current = innerIterator.next();
+                    if (patternMatchPredicate.test(current)) {
+                        return true;
+                    }
+                }
+                if (!innerIterator.hasNext()) { // End of the batch
+                    if (nextBatch) {    // Already fetched the next batch
+                        break;
+                    }
+                    if (results.getObjects().size() < results.getMaxKeys()) {   // This batch was last if less than max keys were returned.
+                        break;
+                    }
+                    // Fetching from the next batch
+                    results = client.listMoreObjects(results);
+                    innerIterator = results.getObjects().iterator();
+                    nextBatch = true;
+                }
+            }
+            results = null;
+            innerIterator = null;
+            current = null;
+            return false;
+        }
+
+        @Override
+        public SegmentProperties next() {
+            if (null != current) {
```

Comment:
This doesn't work as it should. In an `Iterator`, when you invoke `next`, you get the "next" result and on a subsequent invocation, you should get the following item.

Your implementation does not do this. Actually it seems to to advance the "next" every time you invoke `hasNext` ... so if you do: 
```
i.hasNext();
i.hasNext();
```

then your iterator has advanced two positions. `hasNext` should be a side-effect-free operation.